common/account_utils/src/validator_definitions.rs

```diff
@@ -139,6 +152,7 @@ impl ValidatorDefinitions {
             .create_new(false)
             .open(&config_path)
             .map_err(Error::UnableToOpenFile)?;
+        Lockfile::new(config_path.join(LOCK_FILE))?;
```

Comment:
Thanks for taking the time to implement this!

We need to hold onto the `Lockfile` object after it's created, or the [`Drop` implementation](https://github.com/sigp/lighthouse/blob/3b600acdc5bf9726367c18277a22486573b8b457/common/lockfile/src/lib.rs#L66-L70) will delete the file (allowing other processes to create/lock it).

We need to add a lockfile field to the `ValidatorDefinitions` like:

```rust
pub struct ValidatorDefinitions {
    definitions: Vec<ValidatorDefinition>,
    lockfile: Lockfile,
}
```

This will require changing a bit more of the code in this file, as `ValidatorDefinitions` will no longer be able to implement `Deserialize`/`Serialize`. Instead, we can initialise the `definitions` field using the impl for `Vec<ValidatorDefintion>`:

```rust
let definitions = serde_yaml::from_reader(file).map_err(Error::UnableToParseFile)?;
Ok(Self {
    definitions,
    lockfile,
})
```

Similarly for serialisation (we need to skip the `lockfile`).

Finally, it would be great to have a short self-contained test that tries to open the same validation definitions file twice (and fails the 2nd time).

Again, thank you!~ Let me know if you get stuck and I can help you out

Reply:
> Instead, we can initialise the definitions field using the impl for `Vec<ValidatorDefintion>`

As in, wrap the return value in `ValidatorDefinitions::open` (and every other function returning `Self` anywhere in its signature)?

> Similarly for serialisation (we need to skip the `lockfile`).

Assuming `#[serde(skip)]` works on this field, then? I can't think of a scenario where the lockfile type itself should ever end up being represented along with the definitions?

Reply:
> As in, wrap the return value in ValidatorDefinitions::open (and every other function returning Self anywhere in its signature)?

Yep! There might be some other changes required as the change to the struct ripples through, but they'll be similar.

> Assuming #[serde(skip)] works on this field, then? I can't think of a scenario where the lockfile type itself should ever end up being represented along with the definitions?

Yeah, we could use `serde(skip_serializing)`. I think `serde(skip)` would be awkward because of the requirement for a default while deserializing. But see how you go

Reply:
> I think `serde(skip)` would be awkward because of the requirement for a default while deserializing.

Agreed, but then [the `skip_serializing` alternative requires `Lockfile` to be `Deserialize`](https://serde.rs/attr-skip-serializing.html). If the usual `derive` version of this is acceptable then this works fine.

Perhaps I need more perspective; in what contexts will we be likely to deserialise a `Lockfile`?

Reply:
Sorry, I wasn't super clear. To put it another way: I don't think `ValidatorDefinitions` needs to implement `Serialize` _or_ `Deserialize` any more. In the cases where we previously deserialized it, we can just deserialize the single field `definitions: Vec<ValidatorDefinition>`, and in the cases where we previously serialized it we can serialize `Vec<ValidatorDefinition>`. You're right that (de)serializing a lockfile doesn't make any sense.

One way to implement the serialisation skipping the lockfile would be for `ValidatorDefinitions` to implement `Serialize` _but not_ `Deserialize`. Then we could use `serde(skip_serializing)` in isolation. However that's a bit strange, and it's probably better if we're just explicit about operating on just the `definitions` field.

Reply:
@michaelsproul Consult 83fd9da for both the ser/de changes and the ripple effects. Tests are WIP.

Reply:
Looks good! I left a few comments below